### PR TITLE
Update vector hyperdiffusion to act on full vectors

### LINKED ISSUE #1654: Update vector hyperdiffusion to act on full vectors
Currently the vector hyperdiffusion operation in ClimaAtmos does not operate like the standard vector hyperdiffusion. This is because the current vector Laplacian used in the implementation does not act on three-dimensional vector fields.

Loosely speaking, we are currently doing this:
```
scurl = Operators.Curl()
sdiv = Operators.Divergence()
wcurl = Operators.WeakCurl()
wgrad = Operators.WeakGradient()

# typeof(u_h) -> Covariant12Vector
laplacian(u_h) = @. wgrad(sdiv(u_h)) - Covariant12Vector(wcurl(Covariant3Vector(scurl(u_h))))
# typeof(laplacian(u_h)) -> Covariant12Vector
```

We should use the correct vector Laplacian by chaining together differential operators and coordinate transforms that act on three-dimensional vector fields. For example like this:
```
scurl = Operators.Curl()
sdiv = Operators.Divergence()
wcurl = Operators.WeakCurl()
wgrad = Operators.WeakGradient()

# typeof(u_bar) -> Covariant123Vector
laplacian(u_bar) = @. wgrad(sdiv(u_bar)) - Covariant123Vector(wcurl(Covariant123Vector(scurl(u_bar))))
# typeof(laplacian(u_bar)) -> Covariant123Vector

# TODO: 
Interpolate third component to faces.
```
In this pseudocode, the main software issue lies with `scurl(u_bar)` because our spectral element curl operator currently does not operate on `Covariant123Vector` types out of the box. The fix seems reasonably simple in ClimaCore.jl, but a hacky solution also exists.

### PR DESCRIPTION
**Pull request by bischtob**
## Purpose
Closes #1654.

## To-do
- [x] Replace computation of hyperdiffusion on u_h with a computation on a full covariant velocity vector field.
- [x] Check stability of simulations.

I have:
- [x] followed the codebase contribution guide: https://clima.github.io/ClimateMachine.jl/latest/Contributing/
- [x] followed the style guide: https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/
- [x] followed the documentation policy: https://github.com/CliMA/policies/wiki/Documentation-Policy
- [x] checked that this PR does not duplicate an open PR.

In the Content, I have included 
- [x] relevant unit tests, and integration tests, 
- [x] appropriate docstrings on all functions, structs, and modules, and included relevant documentation.

----
- [x] I have read and checked the items on the review checklist.

### CODE REVIEW

File: src/tendencies/hyperdiffusion.jl [hunk 1]
```diff
@@ -60,12 +60,18 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
     end
-
-    @. ᶜ∇²uₕ = C12(wgradₕ(divₕ(Y.c.uₕ)))
+    
+    @. ᶜ∇²ᶜu = C123(wgradₕ(divₕ(p.ᶜu)))
     # Without the C12(), the right-hand side would be a C1 or C2 in 2D space.
     if point_type <: Geometry.Abstract3DPoint
-        # TODO: Are these vector conversions correct when there is topography?
-        @. ᶜ∇²uₕ -= C12(wcurlₕ(C3(curlₕ(Y.c.uₕ))))
+        ᶜu₁₂ = @. Geometry.C12(p.ᶜu)
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:67]:
```suggestion
        ᶜu₁₂ = @. C12(p.ᶜu)
```

File: src/tendencies/hyperdiffusion.jl [hunk 2]
```diff
@@ -60,12 +60,18 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
+        ᶜu₃ = @. Geometry.C3(p.ᶜu)
+        # do the first curl
+        ᶜχ₁₂₃ = @. Geometry.C123(scurl(ᶜu₁₂)) + Geometry.C123(scurl(ᶜu₃))
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:70]:
```suggestion
        ᶜχ₁₂₃ = @. C123(scurl(ᶜu₁₂)) + C123(scurl(ᶜu₃))
```

File: src/tendencies/hyperdiffusion.jl [hunk 3]
```diff
@@ -60,12 +60,18 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
-
-    @. ᶜ∇²uₕ = C12(wgradₕ(divₕ(Y.c.uₕ)))
+    
+    @. ᶜ∇²ᶜu = C123(wgradₕ(divₕ(p.ᶜu)))
     # Without the C12(), the right-hand side would be a C1 or C2 in 2D space.
     if point_type <: Geometry.Abstract3DPoint
-        # TODO: Are these vector conversions correct when there is topography?
-        @. ᶜ∇²uₕ -= C12(wcurlₕ(C3(curlₕ(Y.c.uₕ))))
+        ᶜu₁₂ = @. Geometry.C12(p.ᶜu)
+        ᶜu₃ = @. Geometry.C3(p.ᶜu)
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:68]:
```suggestion
        ᶜu₃ = @. C3(p.ᶜu)
```

File: src/tendencies/hyperdiffusion.jl [hunk 4]
```diff
@@ -60,12 +60,18 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
+        # do the first curl
+        ᶜχ₁₂₃ = @. Geometry.C123(scurl(ᶜu₁₂)) + Geometry.C123(scurl(ᶜu₃))
+        ᶜχ₁₂ = @. Geometry.C12(ᶜχ₁₂₃)
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:71]:
```suggestion
        ᶜχ₁₂ = @. C12(ᶜχ₁₂₃)
```

File: src/tendencies/hyperdiffusion.jl [hunk 5]
```diff
@@ -60,12 +60,18 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
+        ᶜχ₃ = @. Geometry.C3(ᶜχ₁₂₃)
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:72]:
```suggestion
        ᶜχ₃ = @. C3(ᶜχ₁₂₃)
```

File: src/tendencies/hyperdiffusion.jl [hunk 6]
```diff
@@ -60,12 +60,18 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
+        # do the second curl
+        @. ᶜ∇²ᶜu -= Geometry.C123(wcurl(ᶜχ₁₂)) + Geometry.C123(wcurl(ᶜχ₃))
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:74]:
```suggestion
        @. ᶜ∇²ᶜu -= C123(wcurl(ᶜχ₁₂)) + C123(wcurl(ᶜχ₃))
```

File: src/tendencies/hyperdiffusion.jl [hunk 7]
```diff
@@ -104,11 +110,22 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
         end
     end
 
-    @. Yₜ.c.uₕ -= κ₄ * divergence_damping_factor * C12(wgradₕ(divₕ(ᶜ∇²uₕ)))
+    @. ᶜ∇⁴ᶜu = divergence_damping_factor * C123(wgradₕ(divₕ(ᶜ∇²ᶜu)))
+    if point_type <: Geometry.Abstract3DPoint
+        ᶜ∇²ᶜu₁₂ = @. Geometry.C12(ᶜ∇²ᶜu)
+        ᶜ∇²ᶜu₃ = @. Geometry.C3(ᶜ∇²ᶜu)
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:116]:
```suggestion
        ᶜ∇²ᶜu₃ = @. C3(ᶜ∇²ᶜu)
```

File: src/tendencies/hyperdiffusion.jl [hunk 8]
```diff
@@ -104,11 +110,22 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
+        # do the curl
+        ᶜχ₁₂₃ = @. Geometry.C123(scurl(ᶜ∇²ᶜu₁₂)) + Geometry.C123(scurl(ᶜ∇²ᶜu₃))
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:118]:
```suggestion
        ᶜχ₁₂₃ = @. C123(scurl(ᶜ∇²ᶜu₁₂)) + C123(scurl(ᶜ∇²ᶜu₃))
```

File: src/tendencies/hyperdiffusion.jl [hunk 9]
```diff
@@ -104,11 +110,22 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
+        ᶜχ₁₂ = @. Geometry.C12(ᶜχ₁₂₃)
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:119]:
```suggestion
        ᶜχ₁₂ = @. C12(ᶜχ₁₂₃)
```

File: src/tendencies/hyperdiffusion.jl [hunk 10]
```diff
@@ -104,11 +110,22 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
+        ᶜχ₃ = @. Geometry.C3(ᶜχ₁₂₃)
+        # do the second curl
+        @. ᶜ∇⁴ᶜu -= Geometry.C123(wcurl(ᶜχ₁₂)) + Geometry.C123(wcurl(ᶜχ₃))
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:122]:
```suggestion
        @. ᶜ∇⁴ᶜu -= C123(wcurl(ᶜχ₁₂)) + C123(wcurl(ᶜχ₃))
```

File: src/tendencies/hyperdiffusion.jl [hunk 11]
```diff
@@ -104,11 +110,22 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
+    end
+
+    @. Yₜ.c.uₕ -= κ₄ * C12(ᶜ∇⁴ᶜu)
     # Without the C12(), the right-hand side would be a C1 or C2 in 2D space.
     if point_type <: Geometry.Abstract3DPoint
-        # TODO: Are these vector conversions correct when there is topography?
-        @. Yₜ.c.uₕ += κ₄ * C12(wcurlₕ(C3(curlₕ(ᶜ∇²uₕ))))
+        @. Yₜ.c.uₕ -= κ₄ * IC2F(C3(ᶜ∇⁴ᶜu))
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:128]: Return type here is inconsistent with lhs I think; based on above lines `ᶜ∇⁴ᶜu` is `C123`, lhs is `C12`, rhs is `C3`

File: src/tendencies/hyperdiffusion.jl [hunk 12]
```diff
@@ -104,11 +106,15 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
+    @. Yₜ.c.uₕ -= κ₄ * divergence_damping_factor * C12(wgradₕ(divₕ(ᶜ∇²u)))
     # Without the C12(), the right-hand side would be a C1 or C2 in 2D space.
     if point_type <: Geometry.Abstract3DPoint
         # TODO: Are these vector conversions correct when there is topography?
-        @. Yₜ.c.uₕ += κ₄ * C12(wcurlₕ(C3(curlₕ(ᶜ∇²uₕ))))
+        # do the first curl
+        ᶜχ₁₂₃ = @. C123(scurl(C12(ᶜ∇²u))) + C123(scurl(C3(ᶜ∇²u)))
+        # do the second curl
+        @. Yₜ.c.uₕ += C12(wcurl(C12(ᶜχ₁₂₃))) + C123(wcurl(C3(ᶜχ₁₂₃)))
+        #@. Yₜ.f.uᵥ += IC2F(C3(wcurl(C12(ᶜχ₁₂₃))) + C123(wcurl(C3(ᶜχ₁₂₃))))
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:117]:
```suggestion
        # TODO: Revisit uᵥ hyperdiffusion tendency: 
        #@. Yₜ.f.uᵥ += IC2F(C3(wcurl(C12(ᶜχ₁₂₃))) + C123(wcurl(C3(ᶜχ₁₂₃))))
```

**Comment by akshaysridhar** [comment]
```ᶜ∇²u = similar(Y.c, C123{FT}),``` is not a valid argument to the `create_dss_buffer` function either :( ; the vector type support issue comes up here.

**Comment by bischtob** [comment]
Seems like the fix is working. Do we need to be mindful of increasing the cache too much?

**Comment by akshaysridhar** [comment]
> Seems like the fix is working. Do we need to be mindful of increasing the cache too much?

I agree we shouldn't be adding cache vars indiscriminately - I'd first like to see the completed mse outcomes; we can revisit the working solution and clean up the cache in the next few commits.

**Comment by akshaysridhar** [comment]
> 



> > Seems like the fix is working. Do we need to be mindful of increasing the cache too much?
> 
> I agree we shouldn't be adding cache vars indiscriminately - I'd first like to see the completed mse outcomes; we can revisit the working solution and clean up the cache in the next few commits.

Where is `ᶜ∇²specific_tracers` defined ? We're unpacking it in the hyperdiffusion cache but I don't see it in `staggered_nohydrostatic_model` `src` file.

**Comment by bischtob** [comment]
scurl and curl not defined

**Comment by bischtob** [comment]
@akshaysridhar which interpolation method is appropriate? Do we need just interp?

File: src/tendencies/hyperdiffusion.jl [hunk 13]
```diff
@@ -93,8 +95,15 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
-                dss_op!(ᶜ∇²specific_energy, buffer.ᶜ∇²specific_energy)
+                # Need to split the DSS computation here, because our DSS 
+                # operations do not accept Covariant123Vector types
+                @. ᶜ∇²uₕ = C12(ᶜ∇²u)
+                @. ᶜ∇²uᵥ = C3(ᶜ∇²u)
+                dss!(ᶜ∇²uₕ, buffer.ᶜ∇²uₕ)
+                dss!(ᶜ∇²uᵥ, buffer.ᶜ∇²uᵥ)
+                @. ᶜ∇²u = C123(ᶜ∇²uₕ) + C123(ᶜ∇²uᵥ)
+
+                dss!(ᶜ∇²specific_energy, buffer.ᶜ∇²specific_energy)
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:106]:
```suggestion
                dss_op!(ᶜ∇²specific_energy, buffer.ᶜ∇²specific_energy)
```

File: src/tendencies/hyperdiffusion.jl [hunk 14]
```diff
@@ -93,8 +95,15 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
                 Spaces.weighted_dss_internal!,
                 Spaces.weighted_dss_ghost!,
             )
-                dss_op!(ᶜ∇²uₕ, buffer.ᶜ∇²uₕ)
-                dss_op!(ᶜ∇²specific_energy, buffer.ᶜ∇²specific_energy)
+                # Need to split the DSS computation here, because our DSS 
+                # operations do not accept Covariant123Vector types
+                @. ᶜ∇²uₕ = C12(ᶜ∇²u)
+                @. ᶜ∇²uᵥ = C3(ᶜ∇²u)
+                dss!(ᶜ∇²uₕ, buffer.ᶜ∇²uₕ)
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:102]:
```suggestion
                dss_op!(ᶜ∇²uₕ, buffer.ᶜ∇²uₕ)
```

File: src/tendencies/hyperdiffusion.jl [hunk 15]
```diff
@@ -93,8 +95,15 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
+                dss!(ᶜ∇²uᵥ, buffer.ᶜ∇²uᵥ)
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:103]:
```suggestion
                dss_op!(ᶜ∇²uᵥ, buffer.ᶜ∇²uᵥ)
```

File: src/tendencies/hyperdiffusion.jl (hunk 13 above)
Comment (akshaysridhar) [reply src/tendencies/hyperdiffusion.jl:106]: These are just consequences of a recently merged PR, not a major issue.

File: src/tendencies/hyperdiffusion.jl [hunk 16]
```diff
@@ -104,11 +113,14 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
+    @. Yₜ.c.uₕ -= κ₄ * divergence_damping_factor * C12(wgradₕ(divₕ(ᶜ∇²u)))
+    # Without the C123(), the right-hand side would be a C1 or C2 in 2D space.
     if point_type <: Geometry.Abstract3DPoint
-        # TODO: Are these vector conversions correct when there is topography?
-        @. Yₜ.c.uₕ += κ₄ * C12(wcurlₕ(C3(curlₕ(ᶜ∇²uₕ))))
+        ᶜχ₁₂₃ = @. C123(curlₕ(C12(ᶜ∇²u))) + C123(curlₕ(C3(ᶜ∇²u)))
+        @. Yₜ.c.uₕ += C12(wcurlₕ(C12(ᶜχ₁₂₃))) + C12(wcurlₕ(C3(ᶜχ₁₂₃)))
+        # Reuse the buffer variable here so we don't need an extra one
+        @. ᶜ∇²uᵥ = C3(wcurlₕ(C12(ᶜχ₁₂₃))) + C3(wcurlₕ(C3(ᶜχ₁₂₃)))
+        @. Yₜ.f.u₃ += ᶠinterp(ᶜ∇²uᵥ)
```
Comment (akshaysridhar) [inline src/tendencies/hyperdiffusion.jl:123]: `ᶜ∇²uᵥ` is a fourth-order diffusion tendency applied to `Yₜ.f.u₃`; naming might be confusing in the future; I suppose it's a tradeoff between re-using buffer vars or assigning more variables to cache.

Comment (bischtob) [reply src/tendencies/hyperdiffusion.jl:123]: Maybe we need another name then.

Comment (bischtob) [reply src/tendencies/hyperdiffusion.jl:123]: Perhaps we should replace the 2 with a 4

File: src/tendencies/hyperdiffusion.jl [hunk 17]
```diff
@@ -104,11 +115,14 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
 
-    @. Yₜ.c.uₕ -= κ₄ * divergence_damping_factor * C12(wgradₕ(divₕ(ᶜ∇²uₕ)))
-    # Without the C12(), the right-hand side would be a C1 or C2 in 2D space.
+    @. Yₜ.c.uₕ -= κ₄ * divergence_damping_factor * C12(wgradₕ(divₕ(ᶜ∇²u)))
+    # Without the C123(), the right-hand side would be a C1 or C2 in 2D space.
     if point_type <: Geometry.Abstract3DPoint
-        # TODO: Are these vector conversions correct when there is topography?
-        @. Yₜ.c.uₕ += κ₄ * C12(wcurlₕ(C3(curlₕ(ᶜ∇²uₕ))))
+        ᶜχ₁₂₃ = @. C123(curlₕ(C12(ᶜ∇²u))) + C123(curlₕ(C3(ᶜ∇²u)))
+        @. Yₜ.c.uₕ += C12(wcurlₕ(C12(ᶜχ₁₂₃))) + C12(wcurlₕ(C3(ᶜχ₁₂₃)))
```
Comment (szy21) [inline src/tendencies/hyperdiffusion.jl:122]:
```suggestion
        @. Yₜ.c.uₕ += κ₄ * (C12(wcurlₕ(C12(ᶜχ₁₂₃))) + C12(wcurlₕ(C3(ᶜχ₁₂₃))))
```
(I think this is correct?)

File: src/tendencies/hyperdiffusion.jl [hunk 18]
```diff
@@ -104,11 +115,14 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
+        # Reuse the buffer variable here so we don't need an extra one
+        @. ᶜ∇²uᵥ = C3(wcurlₕ(C12(ᶜχ₁₂₃))) + C3(wcurlₕ(C3(ᶜχ₁₂₃)))
+        @. Yₜ.f.u₃ += ᶠwinterp(ᶜJ * ᶜρ, ᶜ∇²uᵥ)
```
Comment (szy21) [inline src/tendencies/hyperdiffusion.jl:125]:
```suggestion
        @. Yₜ.f.u₃ += κ₄ * ᶠwinterp(ᶜJ * ᶜρ, ᶜ∇²uᵥ)
```

**Review by szy21 — APPROVED** [review]

**Comment by szy21** [comment]
We should probably update the hyperdiffusion of subgrid-scale velocities [here](https://github.com/CliMA/ClimaAtmos.jl/blob/5dd92818c5309eb9918c5849dd7ed4a1284d5322/src/tendencies/hyperdiffusion.jl#L80) as well. I'm fine with only changing the grid-scale hyperdiffusion in this PR though, maybe just add a note?

**Comment by akshaysridhar** [comment]
> We should probably update the hyperdiffusion of subgrid-scale velocities [here](https://github.com/CliMA/ClimaAtmos.jl/blob/5dd92818c5309eb9918c5849dd7ed4a1284d5322/src/tendencies/hyperdiffusion.jl#L80) as well. I'm fine with only changing the grid-scale hyperdiffusion in this PR though, maybe just add a note?

https://github.com/CliMA/ClimaAtmos.jl/issues/1695 We can track changes to sgs equations here.

**Comment by bischtob** [comment]
We are extending this issue to include EDMFX updraft velocity components.

**Comment by akshaysridhar** [comment]
Updates to SGS tendencies 
(1) Review `sgs tendency` cache : https://github.com/CliMA/ClimaAtmos.jl/blob/d6be3ecdf43091a6dfbd8a8452d9ac8207279526/src/tendencies/hyperdiffusion.jl#L18
(2) Line 125 in commit f26fc97 : https://github.com/CliMA/ClimaAtmos.jl/pull/1658/files adds a hyperdiffusion contribution to the grid-mean u_{3} velocity expression. 

We want the corresponding expression for the sgs tendencies, computed from 3D vector hyperdiffusion operations. This affects lines
https://github.com/CliMA/ClimaAtmos.jl/blob/d6be3ecdf43091a6dfbd8a8452d9ac8207279526/src/tendencies/hyperdiffusion.jl#LL80C40-L80C78
https://github.com/CliMA/ClimaAtmos.jl/blob/d6be3ecdf43091a6dfbd8a8452d9ac8207279526/src/tendencies/hyperdiffusion.jl#L120, 
where the velocity vector is composed of the 2 horizontal grid-mean components, and the vertical _subgrid-scale_ velocity denoted by `Y.f.sgsʲs.:($$j).u₃.components.data.:1`.

File: src/tendencies/hyperdiffusion.jl [hunk 19]
```diff
@@ -10,21 +10,31 @@ function hyperdiffusion_cache(Y, atmos, do_dss)
     isnothing(atmos.hyperdiff) && return (;)
     FT = eltype(Y)
     n = n_mass_flux_subdomains(atmos.turbconv_model)
+
+    # Grid scale quantities
+    ᶜ∇²u = similar(Y.c, C123{FT})
```
Comment (szy21) [inline src/tendencies/hyperdiffusion.jl:15]: Would it be better to put ᶜ∇²u in gs_quantities?

File: src/tendencies/hyperdiffusion.jl [hunk 20]
```diff
@@ -10,21 +10,31 @@ function hyperdiffusion_cache(Y, atmos, do_dss)
...
     gs_quantities = (;
         ᶜ∇²uₕ = similar(Y.c, C12{FT}),
+        ᶜ∇²uᵥ = similar(Y.c, C3{FT}),
         ᶜ∇²specific_energy = similar(Y.c, FT),
         ᶜ∇²specific_tracers = remove_energy_var.(specific_gs.(Y.c)),
     )
+
+    # Sub-grid scale quantities
+    ᶜ∇²uʲs = n == 0 ? (;) : similar(Y.c, NTuple{n, C123{FT}})
```
Comment (szy21) [inline src/tendencies/hyperdiffusion.jl:24]: Similarly, would it be better to put ᶜ∇²uʲs in sgs_quantities?

File: src/tendencies/hyperdiffusion.jl [hunk 21]
```diff
@@ -93,32 +118,57 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
...
     end
     for j in 1:n
-        @. Yₜ.f.sgsʲs.:($$j).u₃.components.data.:1 -=
-            κ₄ * wdivₕ(gradₕ(ᶠ∇²wʲs.:($$j)))
+        if point_type <: Geometry.Abstract3DPoint
+            ᶜχ₁₂₃ = @. C123(curlₕ(C12(ᶜ∇²uʲs.:($$j)))) +
+               C123(curlₕ(C3(ᶜ∇²uʲs.:($$j))))
+            # Reuse the buffer variable here so we don't need an extra one
+            @. ᶜ∇²uᵥʲs.:($$j) = C3(wcurlₕ(C12(ᶜχ₁₂₃))) + C3(wcurlₕ(C3(ᶜχ₁₂₃)))
+            @. Yₜ.f.sgsʲs.:($$j).u₃ += κ₄ * ᶠwinterp(ᶜJ * ᶜρ, ᶜ∇²uᵥʲs.:($$j))
```
Comment (szy21) [inline src/tendencies/hyperdiffusion.jl:170]: I'll double check with Tapio if we want ᶜρ or ᶜρʲ here, but you don't need to wait for this to merge this PR.

**Review by szy21 — APPROVED** [review]
Looks good, thanks!

File: src/tendencies/hyperdiffusion.jl (hunk 21 above)
Comment (szy21) [reply src/tendencies/hyperdiffusion.jl:179]: We can just use what you have now.

Comment (akshaysridhar) [reply src/tendencies/hyperdiffusion.jl:179]: Thanks @szy21 , there are still some allocation issues which we will work through before another round of ci checks.

**Comment by akshaysridhar** [comment]
When running the perf check locally (see below) the allocations incurred exceed the limit by a factor ~103, which is lower than that in the CI build ( ~316). @charleskawczynski when you can spare some time could you please take a look and see what the cause of this discrepancy might be? What's the best way to identify the alloc issues in this PR? 

**Buildkite / Cluster Output** : Commit [ebc75e6](https://github.com/CliMA/ClimaAtmos.jl/commit/ebc75e6b16f249d32dba326502159e90502d769b)
The relevant CI buildkite job link is [here](https://buildkite.com/clima/climaatmos-ci/builds/9985#018859bc-9f95-4155-a261-0b901e9d70fe)

**Local Build** : Commit  [ebc75e6](https://github.com/CliMA/ClimaAtmos.jl/commit/ebc75e6b16f249d32dba326502159e90502d769b)
```
akshaysridhar@Akshays-MacBook-Air.local /Users/akshaysridhar/Research/Codes/ClimaAtmos.jl [tb/vector_hyperdiff]
% julia --color=yes --project=perf perf/flame.jl --job_id flame_perf_target_edmfx --target_job sphere_baroclinic_wave_rhoe_equilmoist_edmfx
[ Info: Setting up single-process ClimaAtmos run
┌ Info: AtmosModel:
│
│                    `model_config`::`ClimaAtmos.SphericalModel`
│                       `perf_mode`::`ClimaAtmos.PerfStandard`
│                  `moisture_model`::`ClimaAtmos.EquilMoistModel`
│                     `energy_form`::`ClimaAtmos.TotalEnergy`
│                    `precip_model`::`ClimaAtmos.Microphysics0Moment`
│                    `forcing_type`::`ClimaAtmos.HeldSuarezForcing`
│                      `subsidence`::`Nothing`
│                  `radiation_mode`::`ClimaAtmos.RRTMGPInterface.AllSkyRadiationWithClearSkyDiagnostics`
│                          `ls_adv`::`Nothing`
│                   `edmf_coriolis`::`Nothing`
│                  `turbconv_model`::`ClimaAtmos.EDMFX{1, Float32}`
│                  `surface_scheme`::`Nothing`
│     `non_orographic_gravity_wave`::`Nothing`
│         `orographic_gravity_wave`::`Nothing`
│                       `hyperdiff`::`ClimaAtmos.ClimaHyperdiffusion{Float32}`
│                       `vert_diff`::`ClimaAtmos.VerticalDiffusion{false, Float32}`
│                  `viscous_sponge`::`Nothing`
└                 `rayleigh_sponge`::`Nothing`
┌ Info: numerics
│   energy_upwinding = Val{:none}()
│   tracer_upwinding = Val{:none}()
│   density_upwinding = Val{:none}()
│   edmfx_upwinding = Val{:none}()
│   apply_limiter = false
└   bubble = true
┌ Info: Time info:
│   dt = "1secs"
│   t_end = "10secs"
└   floor_n_steps = 10
┌ Info: Topography
└   topography = "NoWarp"
┌ Info: z heights
│   z_mesh.faces =
│    26-element Vector{ClimaCore.Geometry.ZPoint{Float32}}:
│     ZPoint(0.0)
│     ZPoint(499.99963)
│     ZPoint(1012.0728)
│     ZPoint(1537.3689)
│     ZPoint(2077.196)
│     ZPoint(2633.0525)
│     ZPoint(3206.659)
│     ZPoint(3800.0076)
│     ZPoint(4415.418)
│     ZPoint(5055.618)
│     ZPoint(5723.8286)
│     ⋮
│     ZPoint(9653.391)
│     ZPoint(10608.847)
│     ZPoint(11648.46)
│     ZPoint(12792.035)
│     ZPoint(14066.937)
│     ZPoint(15512.436)
│     ZPoint(17187.709)
│     ZPoint(19187.818)
│     ZPoint(21679.867)
│     ZPoint(25000.004)
└     ZPoint(30000.0)
Allocating Y: 12.274104 seconds (84.35 M allocations: 4.392 GiB, 4.29% gc time, 95.86% compilation time)
Allocating cache (p): 42.629137 seconds (316.90 M allocations: 14.829 GiB, 6.34% gc time, 97.55% compilation time: 0% of which was recompilation)
[ Info: Using ODE config: `SSP33ShuOsher`
ode_configuration: 0.765140 seconds (2.30 M allocations: 120.759 MiB, 45.34% gc time, 99.96% compilation time)
get_callbacks: 0.248929 seconds (621.14 k allocations: 34.354 MiB, 99.84% compilation time)
Define ode function: 2.175214 seconds (1.44 M allocations: 123.844 MiB, 99.55% compilation time: 0% of which was recompilation)
args_integrator: 11.099248 seconds (7.43 M allocations: 429.149 MiB, 99.89% compilation time: 0% of which was recompilation)
[ Info: Saving diagnostics to HDF5 file on day 0 second 0
Define integrator: 100.174169 seconds (110.51 M allocations: 5.226 GiB, 1.02% gc time, 42.42% compilation time)
get_integrator: 102.466378 seconds (121.11 M allocations: 5.793 GiB, 0.99% gc time, 43.70% compilation time)
[ Info: running step
[ Info: collect profile
[ Info: collecting allocations
[ Info: testing allocations
  2.562060 seconds (26 allocations: 104.750 KiB)
elapsed time (ns):  2562059792
gc time (ns):       0
bytes allocated:    107264
pool allocs:        23
non-pool GC allocs: 3
minor collections:  0
full collections:   0
[ Info: `allocs (flame_perf_target_edmfx)`: 107264
[ Info: Allocation change (allocs/allocs_limit): 103.13846153846154
Allocations limit: Test Failed at /Users/akshaysridhar/Research/Codes/ClimaAtmos.jl/perf/flame.jl:80
  Expression: allocs ≤ allocs_limit[job_id] * buffer
   Evaluated: 107264 ≤ 1040
Stacktrace:
 [1] macro expansion
   @ /Applications/Julia-1.8.app/Contents/Resources/julia/share/julia/stdlib/v1.8/Test/src/Test.jl:464 [inlined]
 [2] macro expansion
   @ ~/Research/Codes/ClimaAtmos.jl/perf/flame.jl:80 [inlined]
 [3] macro expansion
   @ /Applications/Julia-1.8.app/Contents/Resources/julia/share/julia/stdlib/v1.8/Test/src/Test.jl:1363 [inlined]
 [4] top-level scope
   @ ~/Research/Codes/ClimaAtmos.jl/perf/flame.jl:80
Test Summary:     | Fail  Total  Time
Allocations limit |    1      1  0.3s
ERROR: LoadError: Some tests did not pass: 0 passed, 1 failed, 0 errored, 0 broken.
in expression starting at /Users/akshaysridhar/Research/Codes/ClimaAtmos.jl/perf/flame.jl:79
```

**Comment by akshaysridhar** [comment]
Found the issue, I think - 
```
   #         ᶜ∇²uₕʲs = similar(Y.c, NTuple{n, C12{FT}}),
   #         ᶜ∇²uᵥʲs = similar(Y.c, NTuple{n, C3{FT}}),
```
result in a 103x allocation increase - removing these lines and running collecting allocs from the hyperdiff tendency results in a total of 896 allocs which is `0.86*(current CI max)`.

**Comment by charleskawczynski** [comment]
> Found the issue, I think -

This doesn't impact the allocations in the flame script-- we only track allocations in `OrdinaryDiffEq.step!`, since this is the "hot loop".

File: .buildkite/pipeline.yml [hunk 22]
```diff
@@ -830,13 +830,13 @@ steps:
         command: "julia --color=yes --project=perf perf/flame.jl --job_id flame_perf_target_edmfx --target_job sphere_baroclinic_wave_rhoe_equilmoist_edmfx"
         artifact_paths: "flame_perf_target_edmfx/*"
         agents:
-          slurm_mem: 20GB
+          slurm_mem: 40GB
```
Comment (charleskawczynski) [inline .buildkite/pipeline.yml:833]: Yikes, this is a bit much 😬

Comment (charleskawczynski) [reply .buildkite/pipeline.yml:833]: Ohhhh, wait, @akshaysridhar, try changing the profiler back to a sample rate of 0.01!

Comment (akshaysridhar) [reply .buildkite/pipeline.yml]: > Yikes, this is a bit much 😬

Agreed 😬

File: perf/flame.jl [hunk 23]
```diff
@@ -55,8 +55,8 @@ allocs = @allocated OrdinaryDiffEq.step!(integrator)
 allocs_limit = Dict()
 allocs_limit["flame_perf_target"] = 4320
 allocs_limit["flame_perf_target_tracers"] = 185904
-allocs_limit["flame_perf_target_edmfx"] = 32784336
-allocs_limit["flame_perf_target_edmf"] = 8437841392
+allocs_limit["flame_perf_target_edmfx"] = 33117648
+allocs_limit["flame_perf_target_edmf"] = 8636906992
```
Comment (akshaysridhar) [inline perf/flame.jl:59]: @charleskawczynski I think this PR is ready to be merged with your approval.

Comment (charleskawczynski) [reply perf/flame.jl:59]: Yes! Let’s merge!

**Pull request merged**